### PR TITLE
chmatchdup() now handles non-ascii strings consistently

### LINKED ISSUE #3844: setnames() can't set dt with non-ascii names
(It only occurs on the latest dev version of data.table)

The issue is introduced by https://github.com/Rdatatable/data.table/pull/3784 . The cause is `chmatchdup()` is somehow can't handle non-ASCII strings correctly.

Note, it should be Windows-speicific. 

``` r
library(data.table)

nonascii <- c('中', '文')
data.table:::chmatchdup(
  c(nonascii, nonascii), nonascii
)
#> [1] 1 2 1 2

ascii <- c('a', 'b')
data.table:::chmatchdup(
  c(ascii, ascii), ascii
)
#> [1]  1  2 NA NA

dt <- data.table("中文" = 1, "汉字" = 2)
setnames(dt, colnames(dt), colnames(dt))
#> Error in setnames(dt, colnames(dt), colnames(dt)): Some items of 'old' are duplicated (ambiguous) in column names: [中文, 汉字]
```

<sup>Created on 2019-09-09 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>
<details><summary>Session info</summary>

``` r
devtools::session_info()
#> - Session info ----------------------------------------------------------
#>  setting  value                                              
#>  version  R version 3.4.4 (2018-03-15)                       
#>  os       Windows 7 x64 SP 1                                 
#>  system   x86_64, mingw32                                    
#>  ui       RTerm                                              
#>  language (EN)                                               
#>  collate  Chinese (Simplified)_People's Republic of China.936
#>  ctype    Chinese (Simplified)_People's Republic of China.936
#>  tz       Asia/Taipei                                        
#>  date     2019-09-09                                         
#> 
#> - Packages --------------------------------------------------------------
#>  package     * version    date       lib source        
#>  assertthat    0.2.1      2019-03-21 [1] CRAN (R 3.4.4)
#>  backports     1.1.3      2018-12-14 [1] CRAN (R 3.4.4)
#>  callr         3.2.0      2019-03-15 [1] CRAN (R 3.4.4)
#>  cli           1.1.0      2019-03-19 [1] CRAN (R 3.4.4)
#>  crayon        1.3.4      2017-09-16 [1] CRAN (R 3.4.4)
#>  data.table  * 1.12.3     2019-09-09 [1] local         
#>  desc          1.2.0      2018-05-01 [1] CRAN (R 3.4.4)
#>  devtools      2.0.1      2018-10-26 [1] CRAN (R 3.4.4)
#>  digest        0.6.18     2018-10-10 [1] CRAN (R 3.4.4)
#>  evaluate      0.13       2019-02-12 [1] CRAN (R 3.4.4)
#>  fs            1.2.7      2019-03-19 [1] CRAN (R 3.4.4)
#>  glue          1.3.1      2019-03-12 [1] CRAN (R 3.4.4)
#>  highr         0.8        2019-03-20 [1] CRAN (R 3.4.4)
#>  htmltools     0.3.6.9000 2018-05-07 [1] local         
#>  knitr         1.22       2019-03-08 [1] CRAN (R 3.4.4)
#>  magrittr      1.5        2014-11-22 [1] CRAN (R 3.4.4)
#>  memoise       1.1.0      2017-04-21 [1] CRAN (R 3.4.4)
#>  pkgbuild      1.0.3      2019-03-20 [1] CRAN (R 3.4.4)
#>  pkgload       1.0.2      2018-10-29 [1] CRAN (R 3.4.4)
#>  prettyunits   1.0.2      2015-07-13 [1] CRAN (R 3.4.1)
#>  processx      3.3.0      2019-03-10 [1] CRAN (R 3.4.4)
#>  ps            1.3.0      2018-12-21 [1] CRAN (R 3.4.4)
#>  R6            2.4.0      2019-02-14 [1] CRAN (R 3.4.4)
#>  Rcpp          1.0.1      2019-03-17 [1] CRAN (R 3.4.4)
#>  remotes       2.0.2      2018-10-30 [1] CRAN (R 3.4.4)
#>  rlang         0.3.3      2019-03-29 [1] CRAN (R 3.4.4)
#>  rmarkdown     1.12       2019-03-14 [1] CRAN (R 3.4.4)
#>  rprojroot     1.3-2      2018-01-03 [1] CRAN (R 3.4.3)
#>  sessioninfo   1.1.1      2018-11-05 [1] CRAN (R 3.4.4)
#>  stringi       1.4.3      2019-03-12 [1] CRAN (R 3.4.4)
#>  stringr       1.4.0      2019-02-10 [1] CRAN (R 3.4.4)
#>  testthat      2.1.1.9000 2019-06-22 [1] local         
#>  usethis       1.4.0      2018-08-14 [1] CRAN (R 3.4.4)
#>  withr         2.1.2      2018-03-15 [1] CRAN (R 3.4.4)
#>  xfun          0.6        2019-04-02 [1] CRAN (R 3.4.4)
#>  yaml          2.2.0      2018-07-25 [1] CRAN (R 3.4.4)
#> 
#> [1] D:/app/R_lib/3.1
#> [2] D:/app/R/library
```

</details>

### PR DESCRIPTION
**Pull request by shrektan**
Closes #3844

`chmatchdup()` now handles non-ASCII strings correctly. 

More specifically, `chmatchdup ()` calls the C code `chmatchMain()`, which would fall back to `Rf_match()` when non-ASCII strings were detected. However, the implementation was correct only before the param `bool chmatchdup` got introduced. Since `Rf_match()` can't handle the param `chmatchdup`, it would return undesired results for non-ASCII strings.

This PR will address this issue.

### CODE REVIEW

File: src/chmatch.c [hunk 1]
```diff
@@ -52,19 +44,17 @@ static SEXP chmatchMain(SEXP x, SEXP table, int nomatch, bool chin, bool chmatch
     SET_TRUELENGTH(s,0);   // TODO: do we need to set to zero first (we can rely on R 3.1.0 now)?
   }
   const int tablelen = length(table);
-  const SEXP *td = STRING_PTR(table);
```
Comment (jangorecki) [inline src/chmatch.c]: This allowed to reduce number of STRING_ELT calls below. Cannot we have that back using on utable?

Comment (shrektan) [reply src/chmatch.c]: I've made the change. Now the git diff lines are reduced as well. Thanks.

**Comment by shrektan** [comment]
Note `match_logical()` is no longer used by any function. However, I'm not sure deleting the function is desired or not. So I just leave it there.

File: src/chmatch.c [hunk 2]
```diff
@@ -24,26 +24,19 @@ static SEXP chmatchMain(SEXP x, SEXP table, int nomatch, bool chin, bool chmatch
...
-      // The same string in different encodings (where unknown encoding is different to known, too) are different
-      // CHARSXP pointers; this chmatch relies on pointer equality only. We tried mkChar(CHAR(s)) [ what install() does ]
-      // but this impacted the fastest cases too much. Hence fall back to match() on the first unknown encoding detected
-      // since match() considers the same string in different encodings as equal (but slower). See #2538 and #4818.
-      savetl_end();
-      UNPROTECT(1);
-      return (chin ? match_logical(table, x) : match(table, x, nomatch));
+    if (NEED2UTF8(STRING_ELT(x, i))) {
+      ux = PROTECT(allocVector(STRSXP, xlen)); ++nprotect;
+      for (int i=0; i<xlen; i++) SET_STRING_ELT(ux, i, ENC2UTF8(STRING_ELT(x, i)));
```
Comment (MichaelChirico) [inline src/chmatch.c]: not ideal to rely on scoping here, can change to a different loop index? or refactor not to nest the loops

File: src/chmatch.c [hunk 3]
```diff
@@ -24,26 +24,19 @@ static SEXP chmatchMain(SEXP x, SEXP table, int nomatch, bool chin, bool chmatch
   int *ansd = INTEGER(ans);
   if (!length(table)) { const int val=(chin?0:nomatch), n=LENGTH(x); for (int i=0; i<n; ++i) ansd[i]=val; UNPROTECT(1); return ans; }
   savetl_init();
-  const SEXP *xd = STRING_PTR(x);
   const int xlen = length(x);
+  int nprotect = 1;
+  SEXP ux = x;
   for (int i=0; i<xlen; i++) {
-    SEXP s = xd[i];
-    if (s != NA_STRING && ENC_KNOWN(s) != 64) { // PREV: s != NA_STRING && !ENC_KNOWN(s) - changed to fix for bug #5159. The previous fix
```
Comment (MichaelChirico) [inline src/chmatch.c]: helpful for future readers if you leave these comments and then write your own explaining the shortfalls

File: src/chmatch.c [hunk 4]
```diff
@@ -52,19 +45,18 @@ static SEXP chmatchMain(SEXP x, SEXP table, int nomatch, bool chin, bool chmatch
     SET_TRUELENGTH(s,0);   // TODO: do we need to set to zero first (we can rely on R 3.1.0 now)?
   }
   const int tablelen = length(table);
-  const SEXP *td = STRING_PTR(table);
+  SEXP utable = table;
+  for (int i=0; i<tablelen; i++) {
```
Comment (MichaelChirico) [inline src/chmatch.c]: this seems quite useful, is it something that can be functionalized & put into utils.c for easy access elsewhere?

coerceUtfIfNeeded or something

File: src/chmatch.c (hunk 3 above)
Comment (shrektan) [reply src/chmatch.c]: Thanks, added.

File: src/chmatch.c (hunk 2 above)
Comment (shrektan) [reply src/chmatch.c]: Thanks. Refactored.

File: src/chmatch.c (hunk 4 above)
Comment (shrektan) [reply src/chmatch.c]: Good idea. I've functionalized them.

**Comment by shrektan** [comment]
One question, as of `SEXP *xd = STRING_PTR(coerceUtf8IfNeeded(x));`, do I need to `PROTECT()` or not? I mean xd is pointing to `coerceUtf8IfNeeded(x)`, which will be collected if GC being trigger or R is smart enough to know it's in use and won't collect it?

File: src/utils.c [hunk 5]
```diff
@@ -280,3 +280,25 @@ SEXP islockedR(SEXP DT) {
...
 }
 
+
+// To see if the character vector contains non-ASCII strings not encoded in UTF-8
+bool need2utf8(SEXP x) {
+  const int xlen = length(x);
+  for (int i=0; i<xlen; i++)
+    if (NEED2UTF8(STRING_ELT(x, i))) 
+      return(true);
+    return(false);
```
Comment (jangorecki) [inline src/utils.c]: looks like `return(false)` should be two whitespace less, no? `for` could get `{` for readability

File: src/utils.c [hunk 6]
```diff
@@ -280,3 +280,25 @@ SEXP islockedR(SEXP DT) {
...
+}
+// coerce to UTF-8 encoded string if need2utf8() is true
+SEXP coerceUtf8IfNeeded(SEXP x) { 
+  const bool need = need2utf8(x);
+  if (!need) return(x);
```
Comment (jangorecki) [inline src/utils.c]: please put `return(x)` in new line so code coverage can test it

File: src/utils.c [hunk 7]
```diff
@@ -280,3 +280,25 @@ SEXP islockedR(SEXP DT) {
...
+  const int xlen = length(x);
+  SEXP ans = PROTECT(allocVector(STRSXP, xlen));
+  for (int i=0; i<xlen; i++)
+    SET_STRING_ELT(ans, i, ENC2UTF8(STRING_ELT(x, i)));
```
Comment (jangorecki) [inline src/utils.c]: could we take `STRING_ELT(x, i)` out of the loop be defining `STRING_PTR` and re-using it? more readable and likely more efficient too

File: src/utils.c (hunk 5 above)
Comment (shrektan) [reply src/utils.c]: you are right. thanks.

File: src/utils.c (hunk 7 above)
Comment (shrektan) [reply src/utils.c]: sure.

**Pull request merged**